src/producer.rs

```diff
@@ -5,158 +5,92 @@
-        KafkaProducer {
-            producer: producer_config
-                .create_with_context(CaptureErrorContext)
-                .expect("Producer creation error"),
-        }
-    }
-}
+    type Error;
 
-impl Producer for KafkaProducer {
-    fn send(&mut self, topic_name: &str, payload: &[u8]) -> Result<(), ClientError> {
-        let record: BaseRecord<'_, [u8], [u8]> = BaseRecord::to(topic_name).payload(payload);
-        self.producer
-            .send(record)
-            .map_err(|(error, _message)| ClientError::SendFailed(error))
-    }
+    fn send(&mut self, payload: Message) -> Result<(), Self::Error>;
 }
 
 #[cfg(test)]
+#[derive(Debug, Default)]
 pub(crate) struct DummyProducer {
-    pub messages: Rc<RefCell<Vec<(String, Vec<u8>)>>>,
+    pub messages: Vec<Message>,
 }
 
 #[cfg(test)]
 impl Producer for DummyProducer {
-    fn send(&mut self, topic_name: &str, payload: &[u8]) -> Result<(), ClientError> {
-        self.messages
-            .borrow_mut()
-            .push((topic_name.to_string(), payload.to_vec()));
+    type Error = std::convert::Infallible;
+
+    fn send(&mut self, message: Message) -> Result<(), Self::Error> {
+        self.messages.push(message);
         Ok(())
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use super::DummyProducer;
-    use super::KafkaConfig;
-    use super::Producer;
-    use rdkafka::config::ClientConfig as RdKafkaConfig;
-    use std::cell::RefCell;
-    use std::collections::HashMap;
-    use std::rc::Rc;
-
-    #[test]
-    fn test_build_producer_configuration() {
-        let config = KafkaConfig::new_producer_config(HashMap::from([
-            (
-                "bootstrap.servers".to_string(),
-                "localhost:9092".to_string(),
-            ),
-            (
-                "queued.max.messages.kbytes".to_string(),
-                "1000000".to_string(),
-            ),
-        ]));
-
-        let rdkafka_config: RdKafkaConfig = config.into();
-        assert_eq!(
-            rdkafka_config.get("queued.max.messages.kbytes"),
-            Some("1000000")
-        );
-    }
+    use super::*;
 
     #[test]
     fn test_dummy_producer() {
-        let messages = Rc::new(RefCell::new(Vec::new()));
-        let mut producer = DummyProducer {
-            messages: Rc::clone(&messages),
+        let mut producer = DummyProducer::default();
+
+        let message = Message {
+            timestamp: 1337,
+            shared_resource_id: "foo".to_owned(),
+            app_feature: "app_feature".to_owned(),
+            usage_unit: UsageUnit::Bytes,
+            amount: 42,
         };
-        let res = producer.send("topic", "message".as_bytes());
-        assert!(res.is_ok());
-
-        assert_eq!(producer.messages.borrow().len(), 1);
+        producer.send(message).unwrap();
+
+        let Message {
+            timestamp,
+            shared_resource_id,
+            app_feature,
+            usage_unit,
+            amount,
```

Comment:
What was the rationale to move the Message data structure inside the producer?
In the previous design that was in the accountant as that is the class that knows the semantic of the message, while the producer was simply wrapper around a Kafka producer that manages the callback.

Reply:
At some point we're probably going to need a HTTP producer for Relay PoP instances and in that case I would prefer to have a concrete type here which I can also receive on the other side and aggregate (which should have quite an effect with the amount of pops we have).

It also gives the producer a little bit more freedom to choose the means of transportation.

> In the previous design that was in the accountant as that is the class that knows the semantic of the message

That does make sense, we can also just send opaque bytes and maybe provide a helper that allows you to get from bytes to a message again (or feed the accountant with a previously encoded message).

Reply:
I made it a `Vec<u8>`, I changed it to owned since we always have an owned vec anyways and in Relay I actually need ownership because the data will be passed along to a different service/actor.

Reply:
> At some point we're probably going to need a HTTP producer for Relay PoP instances and in that case I would prefer to have a concrete type here which I can also receive on the other side and aggregate 

Ok, this was not part of the scope when the library was created, I think this would change the architecture and it would make it meaningful to move the message representation down one layer as you did at first.
I'd do that later when we start supporting the HTTP implementation so we have a clearer picture of what is needed.